File: android/src/main/java/com/securitysuite/CryptoConfig.java
package com.securitysuite;

import com.facebook.react.bridge.ReadableMap;

/**
 * Cryptographic parameters supplied by the application.
 */
public final class CryptoConfig {
  public final String keyAgreementAlgorithm;
  public final String keyFactoryAlgorithm;
  public final String encryptionKeyAlgorithm;
  public final String hmacKeyAlgorithm;
  public final String cipherTransformation;
  public final int gcmTagLength;
  public final int gcmIvLength;

  private CryptoConfig(
      String keyAgreementAlgorithm,
      String keyFactoryAlgorithm,
      String encryptionKeyAlgorithm,
      String hmacKeyAlgorithm,
      String cipherTransformation,
      int gcmTagLength,
      int gcmIvLength
  ) {
    this.keyAgreementAlgorithm = keyAgreementAlgorithm;
    this.keyFactoryAlgorithm = keyFactoryAlgorithm;
    this.encryptionKeyAlgorithm = encryptionKeyAlgorithm;
    this.hmacKeyAlgorithm = hmacKeyAlgorithm;
    this.cipherTransformation = cipherTransformation;
    this.gcmTagLength = gcmTagLength;
    this.gcmIvLength = gcmIvLength;
  }

  public static CryptoConfig fromReadableMap(ReadableMap options) {
    if (options == null) {
      throw new IllegalArgumentException("Crypto options are required");
    }

    return new CryptoConfig(
        requireString(options, "keyAgreementAlgorithm"),
        requireString(options, "keyFactoryAlgorithm"),
        requireString(options, "encryptionKeyAlgorithm"),
        requireString(options, "hmacKeyAlgorithm"),
        requireString(options, "cipherTransformation"),
        requireInt(options, "gcmTagLength"),
        requireInt(options, "gcmIvLength")
    );
  }

  public CryptoConfig merge(ReadableMap options) {
    return options == null ? this : fromReadableMap(mergeMaps(this, options));
  }

  private static ReadableMap mergeMaps(CryptoConfig base, ReadableMap overrides) {
    com.facebook.react.bridge.WritableMap map = com.facebook.react.bridge.Arguments.createMap();
    map.putString("keyAgreementAlgorithm", base.keyAgreementAlgorithm);
    map.putString("keyFactoryAlgorithm", base.keyFactoryAlgorithm);
    map.putString("encryptionKeyAlgorithm", base.encryptionKeyAlgorithm);
    map.putString("hmacKeyAlgorithm", base.hmacKeyAlgorithm);
    map.putString("cipherTransformation", base.cipherTransformation);
    map.putInt("gcmTagLength", base.gcmTagLength);
    map.putInt("gcmIvLength", base.gcmIvLength);

    if (overrides.hasKey("keyAgreementAlgorithm")) {
      map.putString("keyAgreementAlgorithm", overrides.getString("keyAgreementAlgorithm"));
    }
    if (overrides.hasKey("keyFactoryAlgorithm")) {
      map.putString("keyFactoryAlgorithm", overrides.getString("keyFactoryAlgorithm"));
    }
    if (overrides.hasKey("encryptionKeyAlgorithm")) {
      map.putString("encryptionKeyAlgorithm", overrides.getString("encryptionKeyAlgorithm"));
    }
    if (overrides.hasKey("hmacKeyAlgorithm")) {
      map.putString("hmacKeyAlgorithm", overrides.getString("hmacKeyAlgorithm"));
    }
    if (overrides.hasKey("cipherTransformation")) {
      map.putString("cipherTransformation", overrides.getString("cipherTransformation"));
    }
    if (overrides.hasKey("gcmTagLength")) {
      map.putInt("gcmTagLength", overrides.getInt("gcmTagLength"));
    }
    if (overrides.hasKey("gcmIvLength")) {
      map.putInt("gcmIvLength", overrides.getInt("gcmIvLength"));
    }
    return map;
  }

  private static String requireString(ReadableMap map, String key) {
    if (!map.hasKey(key) || map.getString(key) == null) {
      throw new IllegalArgumentException("Missing required crypto option: " + key);
    }
    String value = map.getString(key).trim();
    if (value.isEmpty()) {
      throw new IllegalArgumentException("Missing required crypto option: " + key);
    }
    return value;
  }

  private static int requireInt(ReadableMap map, String key) {
    if (!map.hasKey(key)) {
      throw new IllegalArgumentException("Missing required crypto option: " + key);
    }
    return map.getInt(key);
  }
}

File: ios/CryptoConfig.swift
import Foundation

@available(iOS 13.0, *)
struct CryptoConfig {
    let keyAgreementAlgorithm: String
    let keyFactoryAlgorithm: String
    let encryptionKeyAlgorithm: String
    let hmacKeyAlgorithm: String
    let cipherTransformation: String
    let gcmTagLength: Int
    let gcmIvLength: Int

    static func from(dictionary: NSDictionary?) throws -> CryptoConfig {
        guard let dictionary = dictionary else {
            throw NSError(domain: "CryptoConfig", code: 0, userInfo: [
                NSLocalizedDescriptionKey:
                    "Crypto options are required. Call SecuritySuite.initialize() before crypto APIs.",
            ])
        }

        return CryptoConfig(
            keyAgreementAlgorithm: try requireString(dictionary, key: "keyAgreementAlgorithm"),
            keyFactoryAlgorithm: try requireString(dictionary, key: "keyFactoryAlgorithm"),
            encryptionKeyAlgorithm: try requireString(dictionary, key: "encryptionKeyAlgorithm"),
            hmacKeyAlgorithm: try requireString(dictionary, key: "hmacKeyAlgorithm"),
            cipherTransformation: try requireString(dictionary, key: "cipherTransformation"),
            gcmTagLength: try requireInt(dictionary, key: "gcmTagLength"),
            gcmIvLength: try requireInt(dictionary, key: "gcmIvLength")
        )
    }

    func merged(with dictionary: NSDictionary?) throws -> CryptoConfig {
        guard let dictionary = dictionary else { return self }
        var merged: [String: Any] = [
            "keyAgreementAlgorithm": keyAgreementAlgorithm,
            "keyFactoryAlgorithm": keyFactoryAlgorithm,
            "encryptionKeyAlgorithm": encryptionKeyAlgorithm,
            "hmacKeyAlgorithm": hmacKeyAlgorithm,
            "cipherTransformation": cipherTransformation,
            "gcmTagLength": gcmTagLength,
            "gcmIvLength": gcmIvLength,
        ]
        for (key, value) in dictionary {
            if let stringKey = key as? String {
                merged[stringKey] = value
            }
        }
        return try CryptoConfig.from(dictionary: merged as NSDictionary)
    }

    private static func requireString(_ dictionary: NSDictionary, key: String) throws -> String {
        guard let value = dictionary[key] as? String else {
            throw NSError(domain: "CryptoConfig", code: 0, userInfo: [
                NSLocalizedDescriptionKey: "Missing required crypto option: \(key)",
            ])
        }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw NSError(domain: "CryptoConfig", code: 0, userInfo: [
                NSLocalizedDescriptionKey: "Missing required crypto option: \(key)",
            ])
        }
        return trimmed
    }

    private static func requireInt(_ dictionary: NSDictionary, key: String) throws -> Int {
        let value = dictionary[key]
        if let intValue = value as? Int {
            return intValue
        }
        if let numberValue = value as? NSNumber {
            return numberValue.intValue
        }
        throw NSError(domain: "CryptoConfig", code: 0, userInfo: [
            NSLocalizedDescriptionKey: "Missing required crypto option: \(key)",
        ])
    }
}
